README.md
ARC GNU Tool Chain
==================

This is the main git repository for the ARC GNU tool chain. It contains just
the scripts required to build the entire tool chain.

This is the master from which versions of the scripts for different releases
are branched.

Prequisites
-----------

You will need a Linux like environment (Cygwin and MinGW environments under
Windows should work as well).

You will need the standard GNU tool chain pre-requisites as documented in the
ARC GNU tool chain user guide or on the
[GCC website](http://gcc.gnu.org/install/prerequisites.html)

Finally you will need to check out the repositories for each of the tool chain
components (its not all one big repository), including the linux repository
for building the tool chain. These should be peers of this toolchain
directory. If you have yet to check any repository out, then the following
should be appropriate for creating a new directory, `arc` with all the
components.

    mkdir arc
    cd arc
    git clone git://github.com/foss-for-synopsys-dwc-arc-processors/cgen.git
    git clone git://github.com/foss-for-synopsys-dwc-arc-processors/binutils.git
    git clone git://github.com/foss-for-synopsys-dwc-arc-processors/gcc.git
    git clone git://github.com/foss-for-synopsys-dwc-arc-processors/gdb.git
    git clone git://github.com/foss-for-synopsys-dwc-arc-processors/newlib.git
    git clone git://github.com/foss-for-synopsys-dwc-arc-processors/uClibc.git
    git clone git://github.com/foss-for-synopsys-dwc-arc-processors/linux.git
    git clone git://github.com/foss-for-synopsys-dwc-arc-processors/toolchain.git
    cd toolchain

Building the tool chain
-----------------------

The script `build-all.sh` will build and install both *arc-elf32-* and
*arc-linux-uclibc-* tool chains. The comments at the head of this script
explain how it works and the parameters to use. It uses script
`symlink-trunks.sh` to build a unified source directory.

The script `arc-versions.sh` specifies the branches to use in each component
git repository. It should be edited to change the default branches if
required.

Having built a unified source directory and checked out the correct branches,
`build-all.sh` in turn uses `build-elf32.sh` and `build-uclibc.sh`. These
build respectively the *arc-elf32* and *arc-linux-uclibc* tool chains. Details
of the operation are provided as comments in each script file. Both these
scripts use a common initialization script, `arc-init.sh`.



ChangeLog
2013-03-20  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* dejagnu/baseboards/arc-sim-base.exp: Use installed gcc.
	* run-tests.sh: Enable ELF32 tests
	* run-elf32-tests.sh: Update to just test GDB.

2013-03-20  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* build-elf32.sh: Build GDB.

2013-03-19  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* dejagnu/ftp-extra.exp (ftp_download): Silently ignore requests
	for files > 75 chars long.

2013-03-15  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* dejagnu/telnet-extra.exp: Improve commenting throughout
	(telnet_close): Ensure we have a spawn_id for actions, unset
	spawn_id on close, declare spawn_id global.
	(telnet_exec): Clearer handling of names and args, declare
	spawn_id global.
	(telnet_spawn_open): New proc, declare spawn_id global.
	(telnet_spawn): Rewritten to use its own spawn_id, independent of
	telnet_exec, declare spawn_id global.

2013-03-08  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* dejagnu/baseboards/arc-linux-generic.exp: Set up to read a
	command file on GDB startup.
	* dejagnu/telnet-extra.exp (telnet_failure_check): Better error
	message.
	* run-uclibc-tests.sh: Create a command file to set sysroot.

2013-03-08  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* build-linux.sh: Turn off ARC_MMU_SASID and ARC_CMN_MMAP by default.

2013-03-04  Simon Cook  <simon.cook@embecosm.com>

	* build-elf32.sh: Add support for building big endian ELF tool chain.
	* build-uclibc.sh: Set build_dir based on endianness of tools.
	* run-elf32-tests.sh: Set test directories based on endianness.
	* run-uclibc-tests.sh: Likewise.

2013-02-27  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* build-uclibc.sh: Add ARC_LEGACY_PTRACE_ABI defintion when
	building gdbserver.

2013-02-27  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* build-all.sh (UNISRC): Updated.

2013-02-27  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* arc-versions.sh: Build with 4.4 components apart from GDB.
	* build-elf32.sh: Build with 4.4 components and no GDB.
	* build-linux.sh: Build with 4.4 tool chain.
	* build-uclibc.sh: Add ARC_LEGACY_PTRACE_ABI defintion when
	building gdbserver. Build 4.4 tool chain with GDB 7.5.1 (from 4.8
	tool chain).
	* run-tests.sh: Only run uClibc tests.
	* run-uclibc-tests.sh: Only run GDB tests.

2013-02-26  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* arc-versions.sh: Optionally do a fetch first, to deal with new
	branches.

2013-02-25  Simon Cook  <simon.cook@embecosm.com>

	* arc-versions.sh: Updated binutils, gcc, gdb and newlib branch
	references

2013-02-21  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* build-elf32.sh: Don't build and install GDB.

2013-02-21  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* dejagnu/baseboards/arc-linux-generic.exp: Explicitly select the
	installed compiler.

2013-02-21  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* dejagnu/baseboards/arc-linux-generic.exp: Move extra file loads
	to after the main config file is loaded. Add gdb-extra.exp. Mark
	as using GDB stub.
	* dejagnu/gdb-extra.exp: Created as placeholder.

2013-02-20  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* dejagnu/baseboards/arc-linux-aa4.exp: Move most gdbserver config
	to arc-linux-generic.exp
	* dejagnu/baseboards/arc-linux-nsim.exp: Likewise.
	* dejagnu/baseboards/arc-linux-generic.exp: Set most gdbserver
	config.

2013-02-20  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* dejagnu/baseboards/arc-linux-nsim.exp: Correct code accidentally
	introduced.

2013-02-20  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* dejagnu/baseboards/arc-linux-aa4.exp: Configure to run with
	gdbserver target if GDB is the tool.
	* dejagnu/baseboards/arc-linux-nsim.exp: Likewise.
	* dejagnu/baseboards/arc-linux-generic.exp: Decide base
	configuration to load based on tool.
	* dejagnu/ftp-extra.exp (ftp_download): Add code to set executable
	bits of transferred file if necessary.
	* dejagnu/telnet-extra.exp (telnet_spawn): New proc.

2013-02-18  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* build-linux.sh: Use -fcommon -mno-sdata when building gdbserver.

2013-02-18  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* build-linux.sh: Build a 4.8 static busybox only.
	* run-uclibc-tests.sh: Build directories corrected for 4.8 release.

2013-02-11  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* arc-versions.sh (uclibc): Switch to 4.8 branch.

2013-02-06  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* arc-init.sh (run_check, save_res): Remove use of bash "function"
	keywords.
	* build-all.sh: Correct comment typo.

2013-02-06  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* build-all.sh: Use "=" not "==" in [] comparisons.

2013-02-06  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* build-all.sh, run-tests.sh: Correct handling of --jobs, --load
	and --single-thread.
	* build-uclibc.sh: Correct typo in gdbserver config.

2013-02-05  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* dejagnu/baseboards/arc-linux-aa4.exp: Correct layout
	* dejagnu/baseboards/arc-linux-nsim.exp: Correct assignment of
	hostname.

2013-02-04  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* dejagnu/baseboards/arc-linux-aa4.exp: Specify hostname.
	* dejagnu/baseboards/arc-linux-generic.exp: Don't specify hostname
	* dejagnu/baseboards/arc-linux-nsim.exp: Set hostname using
	get-ip.sh script.
	* get-ip.sh, ip-avail.txt: Created.

2013-02-04  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* arc-init.sh (run_check): Use '=' when specifying board
	(save_res): Correct tests.

2013-02-04  Jeremy Bennett <jeremy.bennett@embecosm.com>

	* arc-init.sh (run_check): Add board as fourth argument.
	(save_res): Only run if there is a result file, and put all
	results to log.
	* dejagnu/baseboards/arc-linux-generic.exp: Take the
	target_install data from ${target_triplet}.
	* dejagnu/baseboards/arc-sim-base.exp: Likewise.
	* run-elf32-tests.sh: Add board when running tests.
	* run-tests.sh (ARC_TEST_TARGET): Correct typo in default value

run-tests.sh
#!/bin/sh

# Copyright (C) 2012 Synopsys Inc.

# Contributor Jeremy Bennett <jeremy.bennett@embecosm.com>

# This file is a master script for building ARC tool chains.

# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 3 of the License, or (at your option)
# any later version.

# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.

# You should have received a copy of the GNU General Public License along
# with this program.  If not, see <http://www.gnu.org/licenses/>.          

#	SCRIPT TO RUN ARC-ELF32 and ARC-LINUX-UCLIBC REGRESSION TESTS
#       =============================================================

# This script runs the ARC regression tests for the arc-elf32- and/or
# arc-uclibc-linux- tool chains. It is designed to work with the source tree
# as organized in GitHub. The arguments have the followign meaning.

# Invocation Syntax

#     run-tests.sh [--source-dir <source_dir>]  [--target <address>]
#                  [--jobs <count>] [--load <load>][--single-thread]
#                  [--elf32 | --no-elf32] [--uclibc | --no-uclibc]
#                  [--big-endian]

# --source-dir <source_dir>

#     The location of the ARC GNU tools source tree. If not specified, the
#     script will use the value of the ARC_GNU environment variable if
#     available.

#     If this argument is not specified, and the ARC_GNU environment variable
#     is also not set, the script will use the parent of the directory where
#     this script is installed.

# --target <address>

#     The address of the target, either symbolic, or as an IP address. The
#     default is aa4-32.

# --jobs <count>

#     Specify that parallel make should run at most <count> jobs. The default
#     is <count> equal to one more than the number of processor cores shown by
#     /proc/cpuinfo.

# --load <load>

#     Specify that parallel make should not start a new job if the load
#     average exceed <load>. The default is <load> equal to one more than the
#     number of processor cores shown by /proc/cpuinfo.

# --single-thread

#     Equivalent to --jobs 1 --load 1000. Only run one job at a time, but run
#     whatever the load average.

# --elf32 | --no-elf32

#     If specified, run the arc-elf32- tests (default is --elf32).

# --uclibc | --no-uclibc

#     If specified, run the arc-uclibc-linux- tests (default is --uclibc).

# --big-endian

#     If specified, test the big-endian version of the tool chains
#     (i.e. arceb-elf32- and arceb-linux-uclibc-). At present this is only
#     implemented for the Linux tool chain.

# This script exits with zero if every test has passed and with non-zero value
# otherwise.

# ------------------------------------------------------------------------------
# Set default values for some options
ARC_TEST_TARGET=aa4_32
make_load="`(echo processor; cat /proc/cpuinfo 2>/dev/null echo processor) \
           | grep -c processor`"
jobs=${make_load}
load=${make_load}
elf32="--elf32"
uclibc="--uclibc"

# Parse options
until
opt=$1
case ${opt} in
    --source-dir)
	shift
	ARC_GNU=`(cd "$1" && pwd)`
	;;

    --target)
	shift
	ARC_TEST_TARGET=$1
	;;

    --jobs)
	shift
	jobs=$1
	;;

    --load)
	shift
	load=$1
	;;

    --single-thread)
	jobs=1
	load=1000
	;;

    --elf32 | --no-elf32)
	elf32=$1
	;;

    --uclibc | --no-uclibc)
	uclibc=$1
	;;

    --big-endian)
        ARC_ENDIAN=big
        ;;
    ?*)
	echo "Usage: ./run-tests.sh [--source-dir <source_dir>]"
        echo "                      [--target <address>]"
        echo "                      [--elf32 | --no-elf32]"
        echo "                      [--uclibc | --no-uclibc]"
        echo "                      [--big-endian]"
	exit 1
	;;

    *)
	;;
esac
[ "x${opt}" = "x" ]
do
    shift
done


# Default source directory if not already set.
if [ "x${ARC_GNU}" = "x" ]
then
    d=`dirname "$0"`
    ARC_GNU=`(cd "$d/.." && pwd)`
fi

# Little endian is default
if [ "x${ARC_ENDIAN}" = "x" ]
then
    ARC_ENDIAN=little
fi

# Set up logfile and results directories if either does not exist
mkdir -p ${ARC_GNU}/logs
mkdir -p ${ARC_GNU}/results

# Parallelism
PARALLEL="-j ${jobs} -l ${load}"

# Export everything needed by sub-scripts
export ARC_GNU
export ARC_ENDIAN
export PARALLEL
export ARC_TEST_TARGET

status=0

# Run the ELF32 tests
# if [ "${elf32}" = "--elf32" ]
# then
#     if ! "${ARC_GNU}"/toolchain/run-elf32-tests.sh
#     then
#         echo "ERROR: arc-elf32- tests failed to run."
#         status=1
#     fi
# fi

# Run the UCLIBC tests
if [ "${uclibc}" = "--uclibc" ]
then
    if ! "${ARC_GNU}"/toolchain/run-uclibc-tests.sh
    then
        echo "ERROR: arc-linux-uclibc- tests failed to run."
        status=1
    fi
fi

exit ${status}



run-elf32-tests.sh
#!/bin/sh

# Copyright (C) 2012 Synopsys Inc.

# Contributor Jeremy Bennett <jeremy.bennett@embecosm.com>

# This file is a master script for building ARC tool chains.

# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 3 of the License, or (at your option)
# any later version.

# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.

# You should have received a copy of the GNU General Public License along
# with this program.  If not, see <http://www.gnu.org/licenses/>.          

#		   SCRIPT TO RUN ARC-ELF32 REGRESSION TESTS
#                  ========================================

# Usage:

#   ./run-elf32-tests.sh

# Prerequisites (NOT tested for):

#   ARC_GNU environment variable must be the absolute address of the default
#   source directory.

#   ${ARC_GNU}/logs must exist and be writable

#   ${ARC_GNU}/results must exist and be writable

# Result is 0 if successful, 1 otherwise.


# Standard setup
. "${ARC_GNU}"/toolchain/arc-init.sh

# Set up logfile and results directories if either does not exist
mkdir -p ${ARC_GNU}/logs
mkdir -p ${ARC_GNU}/results

# Run ELF32 regression and gather results. Gathering results is a separate
# function because of the variation in the location and number of results
# files for each tool.
export DEJAGNU=${ARC_GNU}/toolchain/site.exp
echo "Running elf32 tests"

# Set the build directories

# Create the ELF log file and results directory
logfile_elf="$(echo "${ARC_GNU}")/logs/elf32-check-$(date -u +%F-%H%M).log"
rm -f "${logfile_elf}"
res_elf="$(echo "${ARC_GNU}")/results/elf32-results-$(date -u +%F-%H%M)"
mkdir ${res_elf}

# Location of some files depends on endianess. For now with ELF this is
# ignored, but this code is a holding position.
if [ "${ARC_ENDIAN}" = "little" ]
then
    target_dir=arc-elf32
    bd_elf=${ARC_GNU}/bd-4.4-elf32
    bd_elf_gdb=${ARC_GNU}/bd-4.8-elf32-gdb
else
    target_dir=arceb-elf32
    bd_elf=${ARC_GNU}/bd-4.8-elf32eb-gdb
    bd_elf_gdb=${ARC_GNU}/bd-4.8-elf32eb-gdb
fi

# The target board to use
board=arc-sim

# Run the tests
status=0
# run_check ${bd_elf}     binutils            "${logfile_elf}" ${board} \
#     || status=1
# save_res  ${bd_elf}     ${res_elf} binutils/binutils     "${logfile_elf}" \
#     || status=1
# run_check ${bd_elf}     gas                 "${logfile_elf}" ${board} \
#     || status=1
# save_res  ${bd_elf}     ${res_elf} gas/testsuite/gas     "${logfile_elf}" \
#     || status=1
# run_check ${bd_elf}     ld                  "${logfile_elf}" ${board} \
#     || status=1
# save_res  ${bd_elf}     ${res_elf} ld/ld                 "${logfile_elf}" \
#     || status=1
# run_check ${bd_elf}     gcc                 "${logfile_elf}" ${board} \
#     || status=1
# save_res  ${bd_elf}     ${res_elf} gcc/testsuite/gcc/gcc "${logfile_elf}" \
#     || status=1
# echo "Testing g++..."
# save_res  ${bd_elf}     ${res_elf} gcc/testsuite/g++/g++ "${logfile_elf}" \
#     || status=1
# libgcc and libgloss tests are currently empty, so nothing to run or save.
# run_check ${bd_elf}     target-libgcc       "${logfile_elf}"
# run_check ${bd_elf}     target-libgloss     "${logfile_elf}"
# run_check ${bd_elf}     target-newlib       "${logfile_elf}" ${board} \
#     || status=1
# save_res  ${bd_elf}     ${res_elf} ${target_dir}/newlib/testsuite/newlib \
#     "${logfile_elf}" || status=1
# run_check ${bd_elf}     target-libstdc++-v3 "${logfile_elf}" ${board} \
#     || status=1
# save_res  ${bd_elf}     ${res_elf} \
#     ${target_dir}/libstdc++-v3/testsuite/libstdc++ "${logfile_elf}" \
#     || status=1
run_check ${bd_elf_gdb} sim                 "${logfile_elf}" ${board} \
    || status=1
save_res  ${bd_elf_gdb} ${res_elf} sim/testsuite/sim     "${logfile_elf}" \
    || status=1
run_check ${bd_elf_gdb} gdb                 "${logfile_elf}" ${board} \
    || status=1
save_res  ${bd_elf_gdb} ${res_elf} gdb/testsuite/gdb     "${logfile_elf}" \
    || status=1

exit ${status}


build-all.sh
#!/bin/sh

# Copyright (C) 2012 Synopsys Inc.

# Contributor Jeremy Bennett <jeremy.bennett@embecosm.com>

# This file is a master script for building ARC tool chains.

# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the Free
# Software Foundation; either version 3 of the License, or (at your option)
# any later version.

# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.

# You should have received a copy of the GNU General Public License along
# with this program.  If not, see <http://www.gnu.org/licenses/>.          

#	  SCRIPT TO BUILD ARC-ELF32 AND ARC-LINUX-UCLIBC TOOL CHAINS
#         ==========================================================

# Invocation Syntax

#     build-all.sh [--source-dir <source_dir>]  [--linux-dir <linux_dir>]
#                  [--build-dir <build_dir>] [--install-dir <install_dir>]
#                  [--symlink-dir <symlink_dir>]
#                  [--auto-pull | --no-auto-pull]
#                  [--auto-checkout | --no-auto-checkout]
#                  [--unisrc | --no-unisrc]
#                  [--elf32 | --no-elf32] [--linux | --no-linux]
#                  [--datestamp-install]
#                  [--comment-install <comment>] [--big-endian]
#                  [--jobs <count>] [--load <load>] [--single-thread]
#                  [--enable-multilib | --disable-multilib]

# This script is a convenience wrapper to build the ARC GNU 4.4 tool
# chains. It utilizes Joern Rennecke's build-elf32.sh script and Bendan
# Kehoe/Jeremy Bennett's build-uclibc.sh script. The arguments have the
# following meanings:

# This version is modified to work with the source tree as organized in
# GitHub.

# --source-dir <source_dir>

#     The location of the ARC GNU tools source tree. If not specified, the
#     script will use the value of the ARC_GNU environment variable if
#     available.

#     If this argument is not specified, and the ARC_GNU environment variable
#     is also not set, the script will use the parent of the directory where
#     this script is installed.

# --linux-dir <linux_dir>

#     The location of the ARC linux source tree. If not specified, the script
#     will use (in order)
#     - the value of the LINUXDIR environment variable
#     - a directory named "linux" in the source directory.

#     If a valid Linux directory cannot be determined, the script will
#     terminate with an error.

#     If an argument is specified with this option, then the arc-versions.sh
#     script will assume it is already checked out on the desired branch (to
#     help Linux developers).

# --build-dir <build_dir>

#     The directory in which the unified source tree will be constructed and
#     in which the build directories will be created. If not sepcified, the
#     script will use the source directory.

# --install-dir <install_dir>

#     The directory in which both tool chains should be installed. If not
#     specified, both will be installed in the INSTALL sub-directory of the
#     source directory.

#     The previous scripts used to install in <base>/elf32 and <base>/uclibc
#     folders, which wasted file space by duplicating common material and also
#     needed 2 entries in PATH for elf32/uclibc tools despite being from the
#     same build.

# --symlink-dir <symlink_dir>

#     If specified, the install directory will be symbolically linked to this
#     directory.

#     For example it may prove useful to install in a directory named with the
#     date and time when the tools were built, and then symbolically link to a
#     directory with a fixed name. By using the symbolic link in the users
#     PATH, the latest version of the tool chain will be used, while older
#     versions of the tool chains remain available under the dated
#     directories.

# --auto-checkout | --no-auto-checkout

#     If specified, a "git checkout" will be done in each component repository
#     to ensure the correct branch is checked out. Default is to checkout.

# --auto-pull | --no-auto-pull

#     If specified, a "git pull" will be done in each component repository
#     after checkout to ensure the latest code is in use. Default is to pull.

# --unisrc | --no-unisrc

#     If --unisrc is specified, rebuild the unified source tree. If
#     --no-unisrc is specified, do not rebuild it. The default is --unisrc.

# --elf32 | --no-elf32

#     If specified, build the arc-elf32- tool chain (default is --elf32).

# --linux | --no-linux

#     If specified, build the arc-uclibc-linux- tool chain (default is
#     --linux).

# --datestamp-install

#     If specified, this will append a date and timestamp to the install
#     directory name. (see the comments under --symlink-dir above for reasons
#     why this might be useful).

# --comment-install <comment>

#     If specified, this will append a user specified string <comment> to the
#     install directory name. This may prove useful if building variants of
#     tool chains.

# --big-endian

#     If specified, build the big-endian version of the tool chains
#     (i.e. arceb-elf32- and arceb-linux-uclibc-). At present this is only
#     implemented for the Linux tool chain.

# --jobs <count>

#     Specify that parallel make should run at most <count> jobs. The default
#     is <count> equal to one more than the number of processor cores shown by
#     /proc/cpuinfo.

# --load <load>

#     Specify that parallel make should not start a new job if the load
#     average exceed <load>. The default is <load> equal to one more than the
#     number of processor cores shown by /proc/cpuinfo.

# --single-thread

#     Equivalent to --jobs 1 --load 1000. Only run one job at a time, but run
#     whatever the load average.

# --disable-multilib | --enable-multilib

#     Use these to control whether mutlilibs should be built. If this argument
#     is not used, then the value of the environment variable,
#     DISABLE_MULTILIB, will be used if set. If it is not set, then the
#     default is to enable multilibs.

# Where directories are specified as arguments, they are relative to the
# current directory, unless specified as absolute names.

# We do not recognize the ARC_GNU_ONLY_CONFIGURE and ARC_GNU_CONTINUE
# environment variables, which were used in previous versions of this
# script. If you are using this script, you need to run the whole thing. If
# you want to redo bits, use the underlying scripts, or go into the relevant
# directories and do it by hand!


# ------------------------------------------------------------------------------
# Unset variables, which if inherited as environment variables from the caller
# could cause us grief.
unset builddir
unset INSTALLDIR
unset SYMLINKDIR
unset ARC_ENDIAN
unset PARALLEL
unset autocheckout
unset autopull
unset datestamp
unset commentstamp
unset jobs
unset load

# In bash we typically write function blah_blah () { }. However Ubuntu default
# /bin/sh -> dash doesn't recognize the "function" keyword. Its exclusion
# seems to work for both
build_pathnm ()
{
    if [ "x" = "x${MSYSTEM}" ]
    then
	# Linux
	if echo $1 | grep -q -e "^/"
	then
	    RESULT=$1		# Absolute directory
	else
	    RESULT=`pwd`/$1	# Relative directory
	fi
    else
	# MinGW/MSYS
	if echo $1 | grep -q -e "^[A-Za-z]:"
	then
	    RESULT=$1		# Absolute directory
	else
	    RESULT=`pwd`\$1	# Relative directory
	fi
    fi
    echo $RESULT
}

# Set defaults for some options
autocheckout="--auto-checkout"
autopull="--auto-pull"
do_unisrc="--unisrc"
elf32="--elf32"
linux="--linux"

# Parse options
until
opt=$1
case ${opt} in
    --source-dir)
	shift
	ARC_GNU=`(cd "$1" && pwd)`
	;;

    --linux-dir)
	shift
	LINUXDIR=`(cd "$1" && pwd)`
	;;

    --build-dir)
	shift
	builddir=`(cd "$1" && pwd)`
	;;

    --install-dir)
	# This is tricky, since the install directory may not yet exist, so we
	# can't simply change to that directory to resolve its absolute
	# name. We resolve this by assuming that directories beginning with
	# "/" or (on MinGW/MSYS) "<char>:" are absolute, and just prepending
	# the current working directory to all other examples.
	shift

	INSTALLDIR=$(build_pathnm $@)
	;;

    --symlink-dir)
	# This has the same problem as --install-dir
	shift
	SYMLINKDIR=$(build_pathnm $@)
	;;

    --auto-checkout | --no-auto-checkout)
	autocheckout=$1
	;;

    --auto-pull | --no-auto-pull)
	autopull=$1
	;;

    --unisrc | --no-unisrc)
	do_unisrc=$1
	;;

    --elf32 | --no-elf32)
	elf32=$1
	;;

    --linux | --no-linux)
	linux=$1
	;;

    --datestamp-install)
	datestamp=-`date -u +%F-%H%M`
	;;

    --comment-install)
	shift
	commentstamp=$1
	;;

    --big-endian)
	ARC_ENDIAN="big"
	;;

    --jobs)
	shift
	jobs=$1
	;;

    --load)
	shift
	load=$1
	;;

    --single-thread)
	jobs=1
	load=1000
	;;

    --disable-multilib|--enable-multilib)
	DISABLE_MULTILIB=$1
	;;

    ?*)
	echo "Unknown argument $1"
	echo
	echo "Usage: ./build-all.sh [--source-dir <source_dir>]"
        echo "                      [--linux-dir <linux_dir>]"
        echo "                      [--build-dir <build_dir>]"
        echo "                      [--install-dir <install_dir>]"
	echo "                      [--symlink-dir <symlink_dir>]"
	echo "                      [--auto-checkout | --no-auto-checkout]"
        echo "                      [--auto-pull | --no-auto-pull]"
        echo "                      [--unisrc | --no-unisrc]"
        echo "                      [--elf32 | --no-elf32]"
        echo "                      [--linux | --no-linux]"
	echo "                      [--datestamp-install]"
	echo "                      [--comment-install <comment>]"
        echo "                      [--jobs <count>] [--load <load>]"
        echo "                      [--single-thread]"
	echo "                      [--enable-multilib | --disable-multilib]"
	exit 1
	;;

    *)
	;;
esac
[ "x${opt}" = "x" ]
do
    shift
done

# Default source directory if not already set
if [ "x${ARC_GNU}" = "x" ]
then
    d=`dirname "$0"`
    ARC_GNU=`(cd "$d/.." && pwd)`
fi

# Default Linux directory if not already set.
if [ "x${LINUXDIR}" = "x" ]
then
    if [ -d "${ARC_GNU}"/linux ]
    then
	LINUXDIR="${ARC_GNU}"/linux
    else
	echo "ERROR: Cannot find Linux sources."
	exit 1
    fi
fi

if [ "x${builddir}" = "x" ]
then
    builddir="${ARC_GNU}"
fi

if [ "x${INSTALLDIR}" = "x" ]
then
    INSTALLDIR="${ARC_GNU}/INSTALL"
fi

if [ "x$datestamp" != "x" ]
then
    INSTALLDIR="${INSTALLDIR}$datestamp"
fi

if [ "x$commentstamp" != "x" ]
then
    INSTALLDIR="$INSTALLDIR-$commentstamp"
fi

# Default endian
if [ "x${ARC_ENDIAN}" = "x" ]
then
    ARC_ENDIAN="little"
fi

# Default multilib usage
if [ "x${DISABLE_MULTILIB}" = "x" ]
then
    DISABLE_MULTILIB=--enable-multilib
fi

# Default parallellism
make_load="`(echo processor; cat /proc/cpuinfo 2>/dev/null echo processor) \
           | grep -c processor`"

if [ "x${jobs}" = "x" ]
then
    jobs=${make_load}
fi

if [ "x${load}" = "x" ]
then
    load=${make_load}
fi

PARALLEL="-j ${jobs} -l ${load}"

# All the things we export to the scripts
export UNISRC=unisrc-4.4-gdb-7.5
export ARC_GNU
export LINUXDIR
export INSTALLDIR
export ARC_ENDIAN
export DISABLE_MULTILIB
export PARALLEL

# Change to the build directory
cd ${builddir}

# Set up a logfile
mkdir -p ${PWD}/logs
logfile="$(echo "${PWD}")/logs/all-build-$(date -u +%F-%H%M).log"
rm -f "${logfile}"

# Checkout the correct branch for each tool
echo "Checking out GIT trees" >> "${logfile}"
echo "======================" >> "${logfile}"

echo "Checking out GIT trees ..."
if ! ${ARC_GNU}/toolchain/arc-versions.sh ${autocheckout} ${autopull} \
      >> ${logfile} 2>&1
then
    echo "ERROR: Failed to checkout GIT versions of tools"
    exit 1
fi

# Make a unified source tree in the build directory. Note that later versions
# override earlier versions with the current symlinking version.
if [ "x${do_unisrc}" = "x--unisrc" ]
then
    echo "Linking unified tree" >> "${logfile}"
    echo "====================" >> "${logfile}"

    echo "Linking unified tree ..."
    component_dirs="gdb newlib gcc binutils"
    # component_dirs="gdb newlib binutils gcc"
    rm -rf ${UNISRC}

    if ! mkdir -p ${UNISRC}
    then
	echo "ERROR: Failed to create ${UNISRC}"
	exit 1
    fi

    if ! ${ARC_GNU}/toolchain/symlink-all.sh ${UNISRC} \
	"${component_dirs}" >> "${logfile}" 2>&1
    then
	echo "ERROR: Failed to symlink ${UNISRC}"
	exit 1
    fi
fi

# Optionally build the arc-elf32- tool chain
if [ "x${elf32}" = "x--elf32" ]
then
    if ! "${ARC_GNU}"/toolchain/build-elf32.sh --force
    then
	echo "ERROR: arc-elf32- tool chain build failed."
	exit 1
    fi
fi

# Optionally build the arc-linux-uclibc- tool chain
if [ "x${linux}" = "x--linux" ]
then
    if ! "${ARC_GNU}"/toolchain/build-uclibc.sh --force
    then
	echo "ERROR: arc-linux-uclibc- tool chain build failed."
	exit 1
    fi
fi

# Link to the defined place. Note the introductory comments about the need to
# specify explicitly the install directory.
if [ "x${SYMLINKDIR}" != "x" ]
then
    rm -f ${SYMLINKDIR}
    ln -s ${INSTALLDIR} ${SYMLINKDIR}
fi
